packages/rate-limit-controller/src/RateLimitController.ts

```diff
@@ -149,13 +160,14 @@ export class RateLimitController<
     }
     this.recordRequest(type, origin);
 
-    const implementation = this.implementations[type].method;
+    const implementation: RateLimitedApis[ApiType]['method'] =
+      this.implementations[type].method;
 
     if (!implementation) {
       throw new Error('Invalid api type');
     }
 
-    return implementation(...args);
+    return implementation(...args) as ReturnType<typeof implementation>;
```

Comment:
Why are we casting to the return type of the same function? Seems odd 😅

Reply:
This and line 137 are both caused by the contradictory typing of `RateLimitedApis[ApiType]['method']`, specifically how it's being used simultaneously as the widest type constraint and as the narrow method subtypes for individual `ApiType`s. 

I'll see if there's a way to reconcile this. Hopefully both type casts can be removed if the underlying issue is resolved.

Reply:
Tentatively addressed here: https://github.com/MetaMask/core/pull/1890#discussion_r1391340917

Reply:
Improved here: e8826efdf87ec1d65122f8809eff07f191d0967b

Reply:
See https://github.com/MetaMask/core/pull/1890#discussion_r1392026749

Reply:
Fixed: e8826efdf87ec1d65122f8809eff07f191d0967b